src/analysis/processing/qgsalgorithmfieldcalculator.cpp

```diff
@@ -0,0 +1,222 @@
+{
+  return QObject::tr( "Calculated" );
+}
+
+QList<int> QgsFieldCalculatorAlgorithm::inputLayerTypes() const
+{
+  return QList<int>() << QgsProcessing::TypeVector;
+}
+
+QgsProcessingFeatureSource::Flag QgsFieldCalculatorAlgorithm::sourceFlags() const
+{
+  return QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks;
+}
+
+void QgsFieldCalculatorAlgorithm::initParameters( const QVariantMap &configuration )
+{
+  Q_UNUSED( configuration );
+
+  std::unique_ptr< QgsProcessingParameterFieldForm > field = qgis::make_unique< QgsProcessingParameterFieldForm> ( QStringLiteral( "FIELD" ), QObject::tr( "Result field" ), QStringLiteral( "INPUT" ) );
+  std::unique_ptr< QgsProcessingParameterExpression > expression = qgis::make_unique< QgsProcessingParameterExpression> ( QStringLiteral( "FORMULA" ), QObject::tr( "Formula" ), QVariant(), QStringLiteral( "INPUT" ), false, true );
+
+  addParameter( field.release() );
+  addParameter( expression.release() );
+}
+
+QgsFields QgsFieldCalculatorAlgorithm::outputFields( const QgsFields & ) const
+{
+  return mFields;
+}
+
+QString QgsFieldCalculatorAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm computes a new vector layer with the same features of the input layer, "
+                      "but with an additional attribute. The values of this new attribute are computed from each "
+                      "feature using a mathematical formula, based on the properties and attributes of the feature." );
+}
+
+QgsFieldCalculatorAlgorithm *QgsFieldCalculatorAlgorithm::createInstance() const
+{
+  return new QgsFieldCalculatorAlgorithm();
+}
+
+
+bool QgsFieldCalculatorAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  std::unique_ptr< QgsProcessingFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+
+  if ( !source )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  QMap<QString, QVariant::Type> fieldTypes(
+  {
+    {"Real", QVariant::Double},
+    {"Integer", QVariant::Int},
+    {"String", QVariant::String},
+    {"Date", QVariant::Date},
+  } );
+
+  // prepare fields
+  QVariantMap fieldDefinitionMap = parameters.value( QStringLiteral( "FIELD" ) ).toMap();
+
+  if ( fieldDefinitionMap.isEmpty() )
+  {
+    fieldDefinitionMap.insert( QStringLiteral( "name" ), parameters.value( QStringLiteral( "FIELD_NAME" ) ).toString() );
+    fieldDefinitionMap.insert( QStringLiteral( "type" ), parameters.value( QStringLiteral( "FIELD_TYPE" ) ).toString() );
+    fieldDefinitionMap.insert( QStringLiteral( "length" ), parameters.value( QStringLiteral( "FIELD_LENGTH" ) ).toInt() );
+    fieldDefinitionMap.insert( QStringLiteral( "precision" ), parameters.value( QStringLiteral( "FIELD_PRECISION" ) ).toInt() );
+  }
+
+  const QString fieldType = fieldDefinitionMap.value( QStringLiteral( "type" ) ).toString();
```

Comment:
Better would be if the type value is a QVariant::Type value directly, to avoid the hardcoded list above (which misses a lot of required values, like time, datetime, long, long long, bool, etc)

Reply:
The current list there is very barebones, I only made it work for the particular algorithm usecase. If we use `QVariant::Type` only, how would we know the type name in the provider. E.g. if we have `QVariant::String`, does it mean `text` or `varchar`?